### PR TITLE
Add SourceAddr support

### PR DESCRIPTION
**Pull request by zetaab**
supports both ipv4 and ipv6.

The format is [::1]:51608 or 127.0.0.1:51701, so it will map golang net/http RemoteAddr field.

### CODE REVIEW

**Comment by jcchavezs** [comment]
Hi @zetaab thanks for opening this. I have a few things to mention here:

1. The [http-wasm ABI](https://github.com/http-wasm/http-wasm/blob/main/content/http-handler-abi.md) does not support this method and usually adding a new method requires discussion in the ABI before jumping into the implementation.
2. *I think* we don't need such method. Instead we should add back the header `x-remote-addr`, having remoteAddr as a first class field for the request is a Golang specific concern and hence it isn't portable (see for example what happens with the field `Host` in https://github.com/http-wasm/http-wasm-host-go/blob/main/handler/nethttp/host.go#L121

I will dig into the assignment of the `req.RemoteAddr`.

**Comment by anuraaga** [comment]
> Instead we should add back the header 

Remote addr is based on TCP connection information, not headers, so it wouldn't make sense to translate it and it should probably be a native abi. X-forwarded headers might be what you are thinking of, some frameworks may merge them into their exposed remote addr accessor, but I don't believe Go does, not sure though. Even in such frameworks I believe they would then always have a "originalIP" type of field.

To maintain being "raw", we would ideally have a ABI to get the true remote addr and ensure any x-forwarded headers that came in the input are preserved in the headers passed to wasm.

**Comment by zetaab** [comment]
the difference between Host and RemoteAddr is that Host is really header in packet level. RemoteAddr is not header.

**Comment by jcchavezs** [comment]
Thanks for the feedback @zetaab and @anuraaga. I was looking into this in golang source code and stumbled upon a couple of things:

1. In CGI it comes from REMOTE_ADDR variable.
2. In TCP it comes from the IP source address and TCP source port so no headers involvement.

That said I think it makes sense to add it to the ABI as @anuraaga suggested. Would you be up to open first a PR into https://github.com/http-wasm/http-wasm/blob/main/content/http-handler-abi.md ?

**Review by anuraaga — APPROVED** [review]

**Comment by jcchavezs** [comment]
@zetaab do you mind adding a test for getSourceAddress in https://github.com/http-wasm/http-wasm-host-go/blob/main/tck/run.go#L44 ? (is that all regarding TCK @anuraaga ?)

**Comment by anuraaga** [comment]
A tck test requires both the host and guest side, which is here

https://github.com/http-wasm/http-wasm-host-go/blob/main/tck/guest/main.go

So we have to merge to both before writing the test. We would update the guest sdk version in that module to the unreleased code when doing so.

So we shouldn't block any merges on tck tests. But it reminded me I think we need a test in handlertest, which is this repo's unit tests.

https://github.com/http-wasm/http-wasm-host-go/blob/main/testing/handlertest/testhandler.go

**Comment by zetaab** [comment]
@anuraaga @jcchavezs tests added. One minor issue is that the whole test pipeline does not listen for ipv6, so I could not add tests for ipv6 case.

File: tck/guest/go.mod [hunk 1]
```diff
@@ -2,4 +2,4 @@ module github.com/anuraaga/http-wasm-tck/guest
 
 go 1.19
 
-require github.com/http-wasm/http-wasm-guest-tinygo v0.1.1
+require github.com/http-wasm/http-wasm-guest-tinygo v0.3.1-0.20231031134125-487a6e2eec5e
```
Comment (zetaab) [inline tck/guest/go.mod:5]: not binding to release, it uses current latest commit

File: tck/guest/main.go [hunk 2]
```diff
@@ -214,6 +216,24 @@ func (h *handler) testReadBody(req api.Request, resp api.Response, expectedBody
 	return true, 0
 }
 
+func (h *handler) testGetSourceAddr(req api.Request, resp api.Response, expectedAddr string) (next bool, reqCtx uint32) {
+	addr := req.GetSourceAddr()
+	raw := strings.Split(addr, ":")
```
Comment (zetaab) [inline tck/guest/main.go:221]: port number is always something random in this case. So its difficult to verify

**Review by anuraaga — APPROVED** [review]
Thanks!

**Pull request merged**